76 multiple instances

chep opened the pull request:
Resolves #76

chep commented:
Check have passed but documentation of function is not up to date, I did not add INSTANCE argument.

ncaq commented on copilot-chat-command.el:
```diff
@@ -235,8 +234,7 @@ Given the code line as background info."
   (save-excursion
-    (goto-char (point-min))
-    (push-mark (point-max) nil t)
+    (mark-whole-buffer)
```
The warning fixed in the following commit has been restored, perhaps due to a conflict resolution failure.
[refactor: don't use mark-whole-buffer · refactor: fix static warn by ncaq · Pull Request #122 · chep/copilot-chat.el](https://github.com/chep/copilot-chat.el/pull/122/commits/96decd4a04556519bb5fb946b3e68a19d0f9e2d3)

chep replied:
Oops. Fixed

ncaq commented on copilot-chat-command.el:
```diff
@@ -866,6 +880,16 @@ Clears model cache from memory and disk, then triggers background fetch."
   nil)
 
+(defun copilot-chat--get-list-buffer-create (instance)
+  (let ((list-buffer (get-buffer-create
```
If the burden is not too strong, it would be better to write a document comment.

ncaq commented on copilot-chat-prompt-mode.el:
```diff
@@ -54,52 +54,59 @@
   :keymap copilot-chat-prompt-mode-map)
 
-(defun copilot-chat--write-buffer(data save &optional buffer)
+(defun copilot-chat--write-buffer(instance data save &optional buffer)
```
It is easier to deal with this now than later, as it is just a matter of putting appropriate line breaks in the document.

ncaq commented on copilot-chat-shell-maker.el:
```diff
@@ -33,41 +33,45 @@
   "Create or retrieve the Copilot Chat shell-maker buffer."
-  (unless (buffer-live-p copilot-chat--buffer)
-    (setq copilot-chat--buffer (copilot-chat--shell)))
-  (let ((tempb (get-buffer-create copilot-chat--shell-maker-temp-buffer))
+  (unless (buffer-live-p (copilot-chat-chat-buffer instance))
+    (setf (copilot-chat-chat-buffer instance) (copilot-chat--shell instance)))
+  (let ( (tempb (if (copilot-chat-shell-maker-tmp-buf instance)
+                  (copilot-chat-shell-maker-tmp-buf instance)
+                  (get-buffer-create (copilot-chat--shell-maker-temp-buffer-name instance))))
          (inhibit-read-only t))
+    (setf (copilot-chat-shell-maker-tmp-buf instance) tempb)
     (with-current-buffer tempb
       (markdown-view-mode))
-    copilot-chat--buffer))
+    (copilot-chat-chat-buffer instance)))
```
I feel that it lacks readability because it needlessly goes through variables, or conversely, the same process is repeated multiple times.
I'm a recipient of GitHub Copilot Chat (o1), but how about the following rewrite?

```emacs-lisp
(defun copilot-chat--shell-maker-get-buffer (instance)
  "Create or retrieve the Copilot Chat shell-maker buffer."
  (unless (buffer-live-p (copilot-chat-chat-buffer instance))
    (setf (copilot-chat-chat-buffer instance)
          (copilot-chat--shell instance)))
  (let ((tempb (or (copilot-chat-shell-maker-tmp-buf instance)
                   (get-buffer-create (copilot-chat--shell-maker-temp-buffer-name instance))))))
    (setf (copilot-chat-shell-maker-tmp-buf instance) tempb)
    (with-current-buffer tempb
      (let ((inhibit-read-only t))
        (markdown-view-mode)))
    (copilot-chat-chat-buffer instance))
```

chep replied:
Should be ok

ncaq commented on copilot-chat-shell-maker.el:
```diff
@@ -33,41 +33,45 @@
 ;; Constants
-(defconst copilot-chat--shell-maker-temp-buffer "*copilot-chat-shell-maker-temp*")
-
+(defconst copilot-chat--shell-maker-temp-buffer
+  "*copilot-chat-shell-maker-temp "
+  "Temporary buffer prefix for Copilot Chat shell-maker.")
```
I think it would be less confusing to remove this variable since it duplicates the data that can be retrieved from `instance` with `copilot-chat-shell-maker-tmp-buf`.

chep replied:
This is only the prefix used to name the buffers. I renamed it.

ncaq reviewed (changes requested):
Mechanical indentation changes to places not related to this time will make me happy when I review them if you can separate the PRs as much as possible because the differences will be too much.
If you are insistent on doing so, I have no choice but to…

chep commented:
I added a filter in `copilot-chat-list-refresh` so it only displays buffers inside instance directory.

chep commented:
I rebased onto master and modified new functions : `copilot-chat-add-buffers` and `copilot-chat-del-buffers`. I also added doc.

chep commented:
> Mechanical indentation changes to places not related to this time will make me happy when I review them if you can separate the PRs as much as possible because the differences will be too much. If you are insistent on doing so, I have no choice but to…

Sorry, it's strange that the indentation differs from yours, as we are using the same function to indent.
Since I moved a large part of the code to fit your new file organization and squashed many commits to rebase, it’s difficult to separate the indentation changes from the code changes.
Sorry, I promise I will be more careful next time.

ncaq commented on copilot-chat-copilot.el:
```diff
@@ -88,145 +92,149 @@
 
 (defun copilot-chat--fetch-models-async ()
   "Fetch models asynchronously in the background."
   (when (copilot-chat--ready-p)
```
Either the merging operation has failed or the function has disappeared.

chep replied:
I was not sure if it is useful to fetch models at each chat creation (now we don't create only one chat anymore). Finally, I restored old function because it does not seems to be expensive.

ncaq commented:
> Sorry, it's strange that the indentation differs from yours, as we are using the same function to indent. Since I moved a large part of the code to fit your new file organization and squashed many commits to rebase, it’s difficult to separate the indentation changes from the code changes. Sorry, I promise I will be more careful next time.

How does this happen?
I wonder.
I don't have any particular settings for indentation in Emacs Lisp, and I thought I was letting Emacs take care of it.
Maybe everyone would be happier if we introduced a formatter that runs on CI and use that as a guideline.
I'm not sure which formatter to introduce.

peterbecich commented:
Thanks for this. Following up on this comment https://github.com/chep/copilot-chat.el/issues/126#issuecomment-2727685579
> You'll be able to link a chat instance to a directory, which should be the root directory of your project.

, is this root directory automatically determined?
If not, can we copy the behavior of Emacs Projectile? https://docs.projectile.mx/projectile/projects.html

> To achieve this it contains a lot of project detection logic and project type specific logic.
> Broadly speaking, Projectile identifies projects like this:
>   Directories that contain the special .projectile file
>   Directories under version control (e.g. a Git repo)
>   Directories that contain some project description file (e.g. a Gemfile for Ruby projects or pom.xml for Java maven-based projects)

A command such as `M-x copilot-chat-project` that opens a project "instance" would be terrific i.m.o.

--------

The equivalent Projectile commands are `projectile-run-shell`, `projectile-find-file`, etc. 

> If you call copilot-chat-display from a buffer displaying a file located in that directory (or one of its subdirectories), the corresponding chat buffer will be displayed.

Excellent, `copilot-chat-display` will match the Projectile behavior then. The Projectile commands take in the current buffer file (or directory) and take action on the project to which that buffer/directory belongs.

chep commented:
> Thanks for this. Following up on this comment [#126 (comment)](https://github.com/chep/copilot-chat.el/issues/126#issuecomment-2727685579)
> 
> > You'll be able to link a chat instance to a directory, which should be the root directory of your project.
> 
> , is this root directory automatically determined? If not, can we copy the behavior of Emacs Projectile? https://docs.projectile.mx/projectile/projects.html
> 
> > To achieve this it contains a lot of project detection logic and project type specific logic.
> > Broadly speaking, Projectile identifies projects like this:
> > Directories that contain the special .projectile file
> > Directories under version control (e.g. a Git repo)
> > Directories that contain some project description file (e.g. a Gemfile for Ruby projects or pom.xml for Java maven-based projects)
> 
> A command such as `M-x copilot-chat-project` that opens a project "instance" would be terrific i.m.o.
> 
> The equivalent Projectile commands are `projectile-run-shell`, `projectile-find-file`, etc.
> 
> > If you call copilot-chat-display from a buffer displaying a file located in that directory (or one of its subdirectories), the corresponding chat buffer will be displayed.
> 
> Excellent, `copilot-chat-display` will match the Projectile behavior then. The Projectile commands take in the current buffer file (or directory) and take action on the project to which that buffer/directory belongs.

The current implementation asks for a directory when calling `copilot-chat-display` if it can’t find an instance matching current dir.
As it is a big code change, I’m currently focus on bugs and code readability. We will be able to add features later (.projectile file, git topdir, …).

ncaq reviewed (changes requested):
I think it would be easier to understand if all warnings issued by GitHub Actions were fixed so it would be harder to miss other warnings, what do you think?
I didn't find any problems with the other parts as far as I went through them myself.

chep commented:
> I think it would be easier to understand if all warnings issued by GitHub Actions were fixed so it would be harder to miss other warnings, what do you think?
> I didn't find any problems with the other parts as far as I went through them myself.

I see all checks green. What am I missing?

ncaq commented:
@chep 

> I see all checks green. What am I missing?

Oh, the display is a little complicated.
There is nothing that shows up as an error, but there is a warning.
It is shown as Annotation on the CHECK screen and as a diff in the Pull Request.

[76 multiple instances by chep · Pull Request #119 · chep/copilot-chat.el](https://github.com/chep/copilot-chat.el/pull/119/checks)

chep commented:
> @chep
> 
> > I see all checks green. What am I missing?
> 
> Oh, the display is a little complicated. There is nothing that shows up as an error, but there is a warning. It is shown as Annotation on the CHECK screen and as a diff in the Pull Request.
> 
> [76 multiple instances by chep · Pull Request #119 · chep/copilot-chat.el](https://github.com/chep/copilot-chat.el/pull/119/checks)

It should be OK now !

ncaq commented on copilot-chat.el:
```diff
@@ -3,7 +3,7 @@
 
 ;; Author: cedric.chepied <cedric.chepied@gmail.com>
-;; Version: 2.0.1
+;; Version: 2.1.0
```
I forgot one thing, it would be better to raise the major version.
You changed the name of the customization variable, it is no longer compatible.

chep replied:
Done !

ncaq reviewed (approved):
Please merge at your own convenience.
